lib/gui/components/safe-webview/safe-webview.js

```diff
@@ -0,0 +1,212 @@
+
+      return react.createElement('webview', {
+        ref: 'webview',
+        src: url.href
+      }, []);
+
+    }
+
+    // We have to return null explicitly, undefined is an error in React.
+    return null;
+  }
+
+  /**
+   * @summary Add the Webview events if there is an element
+   */
+  componentDidMount() {
+
+    // There is no element to add events to if 'shouldLoad' is false.
+    if (this.state.shouldLoad) {
+
+      // Events React is unaware of have to be handled manually
+      this.refs.webview.addEventListener('did-start-loading', this.didStartLoading);
+      this.refs.webview.addEventListener('did-fail-load', this.didFailLoad);
+      this.refs.webview.addEventListener('did-stop-loading', this.didStopLoading);
+      this.refs.webview.addEventListener('did-get-response-details', this.didGetResponseDetails);
+      this.refs.webview.addEventListener('new-window', this.constructor.newWindow);
+      this.refs.webview.addEventListener('console-message', this.constructor.consoleMessage);
+    }
+  }
+
+  /**
+   * @summary Remove the Webview events if there is an element
+   */
+  componentWillUnmount() {
+
+    // There is no element to remove events from if 'shouldLoad' is false.
+    if (this.state.shouldLoad) {
+
+      // Events React is unaware of have to be handled manually
+      this.refs.webview.removeEventListener('did-start-loading', this.didStartLoading);
+      this.refs.webview.removeEventListener('did-fail-load', this.didFailLoad);
+      this.refs.webview.removeEventListener('did-stop-loading', this.didStopLoading);
+      this.refs.webview.removeEventListener('did-get-response-details', this.didGetResponseDetails);
+      this.refs.webview.removeEventListener('new-window', this.constructor.newWindow);
+      this.refs.webview.removeEventListener('console-message', this.constructor.consoleMessage);
+
+      // Reset timeout id
+      this.loadingTimeoutId = null;
+    }
+  }
+
+  /**
+   * @summary Start the loading timeout timer
+   */
+  didStartLoading() {
+    if (this.props.timeout && this.loadingTimeoutId === null) {
+      this.loadingTimeoutId = setTimeout(() => {
+        this.setState({
+          shouldLoad: false
+        });
+
+      // Milliseconds to wait before cancelling the page-load.
+      }, this.props.timeout);
+    }
+  }
+
+  /**
+   * @summary Set the element state to hidden
+   */
+  didFailLoad() {
+    this.setState({
+      shouldLoad: false
+    });
+  }
+
+  /**
+   * @summary Remove the timeout
+   */
+  didStopLoading() {
+
+    // We have loaded, remove the queued timeout action.
+    clearTimeout(this.loadingTimeoutId);
+  }
+
+  /**
+   * @summary Set the element state depending on the HTTP response code
+   * @param {Event} event - Event object
+   */
+  didGetResponseDetails(event) {
+    const HTTP_OK = 200;
+    const HTTP_ERR = 400;
+
+    if (event.httpResponseCode < HTTP_OK || event.httpResponseCode >= HTTP_ERR) {
```

Comment:
Does this imply that the webview _does_ support 3xx redirects? Has that been tested? ;)